### DISCUSSION TITLE
Koishi v3 精品问答集

### DISCUSSION THREAD

**Discussion by idlist** [koishijs/koishi#270] [General]
Shigma 在 Discord 中的精品问答的记录，因为在聊天记录里容易埋没所以单独拎出来一点。

**Comment by idlist** [comment]
Q: `session.getChannel`和`session.database.getChannel`和`ctx.database.getChannel`哪家强？

A: 
```
session.getChannel()          // 推荐使用这个
session.database.getChannel() // 兼容性设计，不推荐
ctx.database.getChannel()     // 与上面效果一致
```

然后 1 的语法跟 23 略有不同，1 包含了一个如果没有则自动添加的逻辑

**Comment by idlist** [comment]
Q: 我想让一个插件判断是否有数据库，如果有则用数据库，没有则用临时变量储存，应该怎么做？

A: 首先不推荐这么做（原因不明）。其次，如果你想这么做的话可以在 `ctx.on('connect')` 中检查 `ctx` 是否有 `ctx.database` 属性来判断是否加载了数据库。

**Reply by ilharp** [reply]
原因不明可还行

**Reply by shigma** [reply]
我也不知道为啥原因不明了（

**Comment by idlist** [comment]
Q: 如果我想给 Koishi 贡献插件，那么插件的配置与依赖 （dependencies） 应该怎么写比较好？

A: 推荐进行以下的设置：

1. 插件的 npm 包命名应当为 `koishi-plugin-xxx` 或 `@xxx/koishi-plugin-yyy` 。
2. 使用 [具名插件](https://koishi.js.org/guide/context.html#%E5%85%B7%E5%90%8D%E6%8F%92%E4%BB%B6) 格式，同时 `name` 应不包含 `koishi-plugin-` 前缀。
3. `package.json` 中的 `peerDependencies` 需要包含 `koishi-core` 及其支持的对应版本。

例如
```js
"koishi-core": ">=3" // 支持大版本大于等于 3 的 Koishi
```
或者
```js
"koishi-core": "^3.0.0" // 支持大版本等于 3 的 Koishi
```
这样的 **已发布 npm 包** 能够被对应版本的插件市场（目前是 `plugin-webui` 的饼）收录，从而增加插件的易用性。

4. 另外，npm 包不为 `deprecated`。`deprecated` 的包会被插件市场忽视。

**Comment by idlist** [comment]
Q: 我觉得 koishi 自己的 help 太丑了，想自己手写 help，怎么办？

A: 先用 `ctx.command('help').dispose()` 弃用自带的 help，然后开始撸你的新 help。

其实你也可以考虑修改 [template](https://koishi.js.org/api/global.html#template-path-params) 来自定义部分 help 的用语，因为总地来说，手写 help 还是很累人的。这部分的定义在 [这里](https://github.com/koishijs/koishi/blob/master/packages/koishi-core/src/help.ts#L195) 。

**Comment by idlist** [comment]
虽然和 Koishi 没有直接关系但还是写在这里：

Q：我现在使用 go-cqhttp 1.0.0-beta.4，我的账号又被风控了，应该怎么办？

A：可以尝试删除同目录下线的 `session.token` 和 `device.json`，并删掉 `config.yml` 中的密码，使用扫码登陆。有可能可以解除风控状态。

**Comment by idlist** [comment]
Q: 只要一回复“帮助”，所有的 Koishi 机器人就都出来了！我该如何让“帮助”诱捕器失效？

A: 帮助是一个默认的 `shortcut`，你可以在能拿到 `App` 的地方进行以下的操作以删掉这个 `shortcut`。

```js
App._shortcuts = App._shortcuts.filter(shortcut => shortcut.name != '帮助')
```

在插件之中，你可以通过 `ctx.app` 访问 `App` 实例。

或者现在你可以试试 [这个插件](https://www.npmjs.com/package/koishi-plugin-limit-help)。

**Reply by ilharp** [reply]
```ts
app._shortcuts = app._shortcuts.filter(shortcut => shortcut.name != '帮助')
```

**Comment by ilharp** [comment]


**Reply by ilharp** [reply]
![image](https://user-images.githubusercontent.com/20179549/135911468-a581edd4-d9bf-42cc-bc07-bde990c86ba6.png)

ERROR_I_AM_TIRED

**Reply by ilharp** [reply]
在此感谢下群友 @XxLittleCxX

**Comment by DDEle** [comment]
Q: 说起来升级，有人给我抄从v3到v4的在数据库里把channelid切开的代码么（ 

A:  
```sql
UPDATE channel SET platform = SUBSTRING(id FROM 1 FOR LOCATE(':', id) - 1) WHERE 1;
UPDATE channel SET id = SUBSTRING(id FROM LOCATE(':', id) + 1) WHERE 1;
ALTER TABLE channel ADD PRIMARY KEY (platform);
```
